src/main/scala/fi/oph/koski/validation/KoskiValidator.scala

```diff
@@ -702,6 +703,115 @@ class KoskiValidator(tutkintoRepository: TutkintoRepository, val koodistoPalvelu
+      case a: AmmatillisenTutkinnonSuoritus if a.valmis => {
+        val yhteistenOsasuoritustenKoodit = a.osasuoritukset.toList.flatten.filter(o => o.arvioitu && yhteiset.contains(o.koulutusmoduuli.tunniste.koodiarvo)).map(_.koulutusmoduuli.tunniste.koodiarvo)
+        HttpStatus.validate(
+          yhteistenOsasuoritustenKoodit.distinct.size == yhteistenOsasuoritustenKoodit.size
+        )(KoskiErrorCategory.badRequest.validation.rakenne.duplikaattiOsasuoritus(
+          s"Suorituksella ${suorituksenTunniste(suoritus)} on useampi yhteinen osasuoritus samalla koodilla"))
+      }
+      case _ => HttpStatus.ok
+    }
+
+    def validateYhteislaajuus(suoritus: Suoritus): HttpStatus = suoritus match {
+      case a: AmmatillisenTutkinnonSuoritus if a.valmis && a.suoritustapa.koodiarvo == "reformi" => {
+
+        val yhteislaajuus = a.osasuoritukset.toList.flatten.filter(o => o.arvioitu && yhteiset.contains(o.koulutusmoduuli.tunniste.koodiarvo))
+          .foldLeft(0.0) { (a, b) => a + (b.koulutusmoduuli.laajuus match {
+            case Some(a) => a.arvo
+            case None => 0.0
+          })
+        }
+        HttpStatus.validate(
+          yhteislaajuus.round == 35
+        )(KoskiErrorCategory.badRequest.validation.laajuudet.osasuoritustenLaajuuksienSumma(
+          s"Tutkinnon yhteisen osuuden suoritusten yhteenlasketun laajuuden tulee olla 35"))
+      }
+      case _ => HttpStatus.ok
+    }
+
+
+    def validateOnOsaAlueita(suoritus: Suoritus): HttpStatus = suoritus match {
+      case a: AmmatillinenPäätasonSuoritus => {
+        val yhteisetOsaSuoritukset = a.osasuoritukset.toList.flatten.filter(o => o.arvioitu && yhteiset.contains(o.koulutusmoduuli.tunniste.koodiarvo))
+
+        val missingOsaalueita = yhteisetOsaSuoritukset.filter(yht => {
+          yht.osasuoritukset.getOrElse(List()).isEmpty
+        })
+
+        val errorKuvaus = missingOsaalueita.foldLeft("")((a, b) => {
+          a + b.koulutusmoduuli.tunniste.koodiarvo
+        })
+
+        HttpStatus.validate(
+          missingOsaalueita.isEmpty
+        )(KoskiErrorCategory.badRequest.validation.rakenne.yhteiselläOsuudellaEiOsasuorituksia(
+          s"Tutkinnon yhteisillä osuuksilla $errorKuvaus ei ole osasuorituksia."))
+      }
+      case _ => HttpStatus.ok
+    }
+
+    def validateYhteistenOsienLaajuus(suoritus: Suoritus): HttpStatus = suoritus match {
+      case a: AmmatillinenPäätasonSuoritus => {
+          val yhteisetOsaSuoritukset = a.osasuoritukset.toList.flatten.filter(o => o.arvioitu && yhteiset.contains(o.koulutusmoduuli.tunniste.koodiarvo))
+
+          val mismatchingLaajuudet = yhteisetOsaSuoritukset.filter(yht => {
+            val yläsuorituksenLaajuus = yht.koulutusmoduuli.laajuus match {
+              case Some(a) => a.arvo
+              case None => 0.0
+            }
+            val alasuoritustenLaajuus = yht.osasuoritukset.toList.flatten.foldLeft(0.0) { (a, b) =>
```

Comment:
Voisiko tämän `foldLeft`:in myös kirjoittaa toisin käyttäen korkeamman tason kombinaattoreita?